sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala

```diff
@@ -281,11 +284,27 @@ case class InsertIntoHiveTable(
             oldPart.flatMap(_.storage.locationUri.map(uri => new Path(uri)))
           }
 
+          val hiveVersion = externalCatalog.asInstanceOf[ExternalCatalogWithListener]
+            .unwrapped.asInstanceOf[HiveExternalCatalog]
+            .client
+            .version
+          // https://issues.apache.org/jira/browse/SPARK-31684,
+          // For Hive 2.0.0 and onwards, as https://issues.apache.org/jira/browse/HIVE-11940
+          // has been fixed, and there is no performance issue anymore.
```

Comment:
Just to check; have you checked that Spark(w/ this fix)+Hive2.0.0+ has the same performance with the current Spark (w/o this fix)?

Reply:
I haven't done the perf tests. According to the discussion in https://github.com/apache/spark/pull/15726, I choose to believe at the moment

Reply:
This code block was there for a long time, would be better to re-run benchmark now, to make sure we don't introduce regressio.

Reply:
Any existing benchmark？Or I can try to add one.

Reply:
yea please add one.

Reply:
No problem.

Reply:
I have just added a benchmark for InsertIntoHiveTable and updated the results.
It uses the "INSERT INTO" as the control group, and "INSERT OVERWRITE" as the experimental group.
With builtin hive 2.3.7, the results of 2 groups are close.
WIth builtin hive 1.2.1.xxx, the results of the experimental group reveal a huge performance degradation when the dynamic partition column exists (the reason here is that #15726 is not merged.